Fast nixos test eval

roberth opened the pull request:
###### Description of changes

Speed up `nixosTests` evaluation by reusing nixpkgs.

 - Add a module that makes `nixpkgs.*` options read-only (except `nixpkgs.pkgs` which is set-once)
 - Add a module for nixpkgs sharing in `all-tests.nix`. Use the preceding module to make sure we don't test the wrong thing.
 - Some minor improvements; see commit list.

As for the performance improvement in acme: about **90 seconds** down to **40 seconds**

Heap stats

Vanilla: **5.2G** final heap, **9.5G** total allocations

```
  "gc": {
    "heapSize": 5235208192,
    "totalBytes": 9486783456
  }
```

Read only **2.3G** final heap, **5.3G** total allocations

```
  "gc": {
    "heapSize": 2299195392,
    "totalBytes": 5271224720
  }
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
lib.types.pkgs: init

A nominal type.

roberth committed:
nixos/nixpkgs: Use types.pkgs

roberth committed:
nixos/eval-config: Remove statically known mkIf

mkIf is unnecessary when the condition is statically known - that is
knowable before entering the module evaluation.

By changing this to a precomputed module, we support changing the
defined options to readOnly options.

roberth committed:
nixosModules.pkgsReadOnly: init

roberth committed:
nixos/all-tests.nix: Set nixpkgs.system

roberth committed:
nixos/all-tests.nix: Add readOnlyPkgs module

roberth committed:
nixosTests.acme: Use a read-only pkgs

This speeds up evaluation by a factor 2.

Ballpark figures from my machine:

```
$ time nix-build nixos/release.nix -A tests.acme
/nix/store/q4fxp55k64clcarsx8xc8f6s10szlfvz-vm-test-run-acme
/nix/store/lnfqg051sxx05hclva84bcbnjfc71c8x-vm-test-run-acme

real    1m28.142s
user    1m7.474s
sys     0m7.932s

$ time nix-build nixos/release.nix -A tests.acme
/nix/store/q4fxp55k64clcarsx8xc8f6s10szlfvz-vm-test-run-acme
/nix/store/lnfqg051sxx05hclva84bcbnjfc71c8x-vm-test-run-acme

real    0m38.235s
user    0m33.814s
sys     0m2.283s

```

Kranzes commented on nixos/tests/all-tests.nix:
```diff
@@ -77,7 +98,7 @@ in {
   _3proxy = runTest ./3proxy.nix;
   aaaaxy = runTest ./aaaaxy.nix;
-  acme = runTest ./acme.nix;
+  acme = runTest { imports = [ ./acme.nix readOnlyPkgs ]; };
```
Is the plan to set this for all tests?

roberth replied:
I think that would be best, but I can't speak for everyone.
Most tests aren't in `runTest` style yet, and I don't know if everyone wants to switch their test to it.
Doing it for `handleTest` is too complicated for me. Too many callbacks and alternate ways to invoke stuff.

Kranzes replied:
We can default it for runTest for now, right?

roberth replied:
Some modules will set something in `nixpkgs.config`, or add an overlay. Wouldn't recommend, but that means that we can't apply this everywhere. I don't mind if "leaky" things like this are repeated. It's good to be reminded of it.

roberth replied:
I've added a `node.pkgsReadOnly` option to the test framework so that the optimization can be disabled if needed.
We could run it through the `python-test-refactoring` job on hydra.nixos.org to figure out which tests need an opt out. https://nixos.org/manual/nixos/unstable/index.html#sec-test-the-test-framework
Although that's not necessary for this PR.

roberth committed:
nixos/testing: Add node.pkgs option

By factoring out this logic, it's easier for other projects to make
use of it this optimization too (and do it correctly).

roberth committed:
nixos/testing: Add node.pkgsReadOnly escape hatch

By adding this option indirection, a test can declare all by itself
that it needs a custom nixpkgs. This is a more convenient way of
going about this when the caller of the test framework receives a
`node.pkgs` unconditionally.

roberth committed:
nixos/testing/nodes.nix: Do not rely on disabledModules

It's just not necessary.

roberth committed:
nixos/all-tests: Enable readOnlyPkgs by default for runTest

Most tests are not affected by this because they use the `handleTest`
function instead.

tfc commented:
Looks good to me. It would be great to have it run through all the tests to find out which one needs an opt-out to make the majority of them faster.

roberth commented:
I've removed the change that disabled aliases on hydra, because a few tests were not compatible with that.

Other than that, I've added this to `all-tests.nix`'s `runTest` and tested on hydra that it works.
For now it has a small effect though, because many tests still use the `handleTest`, as [mentioned](https://github.com/NixOS/nixpkgs/pull/230523#discussion_r1186867922).

Kranzes commented:
Does this optimization also works by default on out-of-tree tests that use `nixos-lib.runTest`?

roberth commented:
> Does this optimization also works by default on out-of-tree tests that use `nixos-lib.runTest`?

Yeah, just set `node.pkgs = pkgs;`

That reminds me, https://github.com/NixOS/nixpkgs/pull/225313 should use it.

roberth commented:
Individual test are not worth reporting on, but we could announce that this is happening.

> warning / deprecation notice

They're not really a public interface, in the sense that nixpkgs is a library of packages and modules and whatnot, but it is a frequently used user interface nonetheless.

Speaking of policy, this should probably be a formal community decision before we migrate other people's tests. Do you think it should be an RFC?

RaitoBezarius commented:
> Individual test are not worth reporting on, but we could announce that this is happening.

Individual no, general notice yes. (with limited examples) :)

> 
> > warning / deprecation notice
> 
> They're not really a public interface, in the sense that nixpkgs is a library of packages and modules and whatnot, but it is a frequently used user interface nonetheless.

That's true, but we are big enough to make it nice for people working with internal APIs to be noticed about changes I believe. People who does this work are the most likely to read the changelog properly, so it's extra appreciated I believe.

> Speaking of policy, this should probably be a formal community decision before we migrate other people's tests. Do you think it should be an RFC?

I don't believe it should be a RFC, test ownership is already inferior to module ownership which is itself inferior to package ownership.

But we should definitely ping maintainers on each of their tests when they are listed as maintainers, but I would not block any decision on them.

roberth commented:
Fair enough. If it's individual / collaborative choice and not policy, it's still worth noting.
